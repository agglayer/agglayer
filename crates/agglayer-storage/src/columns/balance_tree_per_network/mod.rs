use super::{ColumnSchema, BALANCE_TREE_PER_NETWORK_CF};

#[cfg(test)]
mod tests;

/// Column family for the balance tree per network.
///
/// ## Column definition
///
/// | key                               | value                                   |
/// | ---                               | --                                      |
/// | (`NetworkId`, `SmtKeyType::Root`) | (`hash(root.left)`, `hash(root.right)`) |
/// | (`NetworkId`, `hash(node)`)       | (`hash(node.left)`, `hash(node.right)`) |
/// | (`NetworkId`, `hash(leaf)`)       | (`hash(leaf)`)                          |
pub struct BalanceTreePerNetworkColumn;

impl ColumnSchema for BalanceTreePerNetworkColumn {
    type Key = crate::types::SmtKey;
    type Value = crate::types::SmtValue;

    const COLUMN_FAMILY_NAME: &'static str = BALANCE_TREE_PER_NETWORK_CF;
}
