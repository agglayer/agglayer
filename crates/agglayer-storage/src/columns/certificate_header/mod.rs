use agglayer_types::CertificateHeader;

use super::{ColumnSchema, CERTIFICATE_HEADER_CF};

#[cfg(test)]
mod tests;

/// Column family for the certificate headers.
///
/// ## Column definition
///
/// | key               | value                |
/// | --                | --                   |
/// | `CertificateId`   |  `CertificateHeader` |
pub struct CertificateHeaderColumn;

pub type Key = agglayer_types::CertificateId;
pub type Value = CertificateHeader;

impl ColumnSchema for CertificateHeaderColumn {
    type Key = Key;
    type Value = Value;

    const COLUMN_FAMILY_NAME: &'static str = CERTIFICATE_HEADER_CF;
}
