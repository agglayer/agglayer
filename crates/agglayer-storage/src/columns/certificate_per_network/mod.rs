use agglayer_types::CertificateId;
use serde::{Deserialize, Serialize};

use super::{Codec, ColumnSchema, CERTIFICATE_PER_NETWORK_CF};

#[cfg(test)]
mod tests;

/// Column family for the certificate per network per height.
///
/// ## Column definition
///
/// | key                     | value           |
/// | --                      | --              |
/// | (`NetworkId`, `Height`) | `CertificateId` |
pub struct CertificatePerNetworkColumn;

#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct Key {
    pub(crate) network_id: u32,
    pub(crate) height: u64,
}

impl Codec for Key {}

impl ColumnSchema for CertificatePerNetworkColumn {
    type Key = Key;
    type Value = CertificateId;

    const COLUMN_FAMILY_NAME: &'static str = CERTIFICATE_PER_NETWORK_CF;
}
