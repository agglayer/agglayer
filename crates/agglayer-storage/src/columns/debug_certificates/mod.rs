use agglayer_types::{Certificate, CertificateId};

use super::{ColumnSchema, DEBUG_CERTIFICATES_CF};

/// Column family containing the certificates received.
///
/// ## Column definition
///
/// | key             | value           |
/// | --              | --              |
/// | `CertificateId` | `Certificate`   |
pub(crate) struct DebugCertificatesColumn;

impl ColumnSchema for DebugCertificatesColumn {
    type Key = CertificateId;
    type Value = Certificate;

    const COLUMN_FAMILY_NAME: &'static str = DEBUG_CERTIFICATES_CF;
}
