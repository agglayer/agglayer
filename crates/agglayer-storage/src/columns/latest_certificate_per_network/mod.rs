use serde::{Deserialize, Serialize};

use super::{Codec, ColumnSchema, LATEST_SETTLED_CERTIFICATE_PER_NETWORK_CF};
use crate::types::NetworkId;

#[cfg(test)]
mod tests;

/// Column family for the latest settled certificate per network.
/// The key is the network_id and the value is the certificateID,
/// the height and the epoch_number.
///
/// ## Column definition
///
/// | key         | value                                |
/// | --          | --                                   |
/// | `NetworkId` | (`Proof`, `CertificateId`, `Height`) |
pub struct LatestSettledCertificatePerNetworkColumn;

#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct ProvenCertificate(pub [u8; 32], pub u64, pub u64);

pub type Key = NetworkId;

impl Codec for ProvenCertificate {}

impl ColumnSchema for LatestSettledCertificatePerNetworkColumn {
    type Key = Key;
    type Value = ProvenCertificate;

    const COLUMN_FAMILY_NAME: &'static str = LATEST_SETTLED_CERTIFICATE_PER_NETWORK_CF;
}
