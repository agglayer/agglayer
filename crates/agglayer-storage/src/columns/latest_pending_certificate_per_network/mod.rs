use agglayer_types::{CertificateId, Height, NetworkId};
use serde::{Deserialize, Serialize};

use super::{Codec, ColumnSchema, LATEST_PENDING_CERTIFICATE_PER_NETWORK_CF};

/// Column family for the latest pending certificate per network.
/// The key is the network_id and the value is the certificateID and
/// the height.
///
/// ## Column definition
///
/// | key         | value                       |
/// | --          | --                          |
/// | `NetworkId` | (`CertificateId`, `Height`) |
pub struct LatestPendingCertificatePerNetworkColumn;

#[derive(Clone, Copy, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct PendingCertificate(pub CertificateId, pub Height);

pub type Key = NetworkId;

impl Codec for PendingCertificate {}

impl ColumnSchema for LatestPendingCertificatePerNetworkColumn {
    type Key = Key;
    type Value = PendingCertificate;

    const COLUMN_FAMILY_NAME: &'static str = LATEST_PENDING_CERTIFICATE_PER_NETWORK_CF;
}
