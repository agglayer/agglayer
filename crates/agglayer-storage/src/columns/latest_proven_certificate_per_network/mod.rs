use agglayer_types::{CertificateId, Height, NetworkId};
use serde::{Deserialize, Serialize};

use super::{Codec, ColumnSchema, LATEST_PROVEN_CERTIFICATE_PER_NETWORK_CF};

/// Column family for the latest proven certificate per network.
/// The key is the network_id and the value is the certificateID and
/// the height.
///
/// ## Column definition
///
/// | key         | value                       |
/// | --          | --                          |
/// | `NetworkId` | (`CertificateId`, `Height`) |
pub struct LatestProvenCertificatePerNetworkColumn;

#[derive(Clone, Copy, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct ProvenCertificate(pub CertificateId, pub NetworkId, pub Height);

pub type Key = NetworkId;

impl Codec for ProvenCertificate {}

impl ColumnSchema for LatestProvenCertificatePerNetworkColumn {
    type Key = Key;
    type Value = ProvenCertificate;

    const COLUMN_FAMILY_NAME: &'static str = LATEST_PROVEN_CERTIFICATE_PER_NETWORK_CF;
}
