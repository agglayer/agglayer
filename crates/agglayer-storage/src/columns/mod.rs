use bincode::{DefaultOptions, Options};
use serde::{de::DeserializeOwned, Serialize};

use crate::error::Error;

pub fn default_bincode_options() -> impl bincode::Options {
    DefaultOptions::new()
        .with_big_endian()
        .with_fixint_encoding()
}

// State related CFs
pub const CERTIFICATE_PER_NETWORK_CF: &str = "certificate_per_network_cf";
pub const NULLIFIER_TREE_PER_NETWORK_CF: &str = "nullifier_tree_per_network_cf";
pub const BALANCE_TREE_PER_NETWORK_CF: &str = "balance_tree_per_network_cf";
pub const LOCAL_EXIT_TREE_PER_NETWORK_CF: &str = "local_exit_tree_per_network_cf";

// Metadata CFs
pub const CERTIFICATE_HEADER_CF: &str = "certificate_header_cf";
pub const LATEST_PROVEN_CERTIFICATE_PER_NETWORK_CF: &str =
    "latest_proven_certificate_per_network_cf";
pub const LATEST_SETTLED_CERTIFICATE_PER_NETWORK_CF: &str =
    "latest_settled_certificate_per_network_cf";
pub const LATEST_PENDING_CERTIFICATE_PER_NETWORK_CF: &str =
    "latest_pending_certificate_per_network_cf";
pub const METADATA_CF: &str = "metadata_cf";

// epochs related CFs
pub const PER_EPOCH_CERTIFICATES_CF: &str = "per_epoch_certificates_cf";
pub const PER_EPOCH_METADATA_CF: &str = "per_epoch_metadata_cf";
pub const PER_EPOCH_PROOFS_CF: &str = "per_epoch_proofs_cf";
pub const PER_EPOCH_END_CHECKPOINT_CF: &str = "per_epoch_end_checkpoint_cf";
pub const PER_EPOCH_START_CHECKPOINT_CF: &str = "per_epoch_start_checkpoint_cf";
pub const PER_EPOCH_TRANSACTION_HASH_PER_CERTIFICATE_INDEX: &str =
    "per_epoch_transaction_hash_per_certificate_index";

// Pending related CFs
pub const PENDING_QUEUE_CF: &str = "pending_queue_cf";
pub const PROOF_PER_CERTIFICATE_CF: &str = "proof_per_certificate_cf";

// debug CFs
pub const DEBUG_CERTIFICATES_CF: &str = "debug_certificates";

pub trait Codec: Sized + Serialize + DeserializeOwned {
    fn encode(&self) -> Result<Vec<u8>, Error> {
        Ok(default_bincode_options().serialize(self)?)
    }

    fn decode(buf: &[u8]) -> Result<Self, Error> {
        Ok(default_bincode_options().deserialize(buf)?)
    }
}

pub trait ColumnSchema {
    type Key: Codec;
    type Value: Codec;

    const COLUMN_FAMILY_NAME: &'static str;
}

// State
pub(crate) mod balance_tree_per_network;
pub(crate) mod certificate_per_network;
pub(crate) mod local_exit_tree_per_network;
pub(crate) mod nullifier_tree_per_network;

// Pending
pub(crate) mod pending_queue;
pub(crate) mod proof_per_certificate;

// Metadata
pub(crate) mod certificate_header;
pub mod latest_pending_certificate_per_network;
pub mod latest_proven_certificate_per_network;
pub mod latest_settled_certificate_per_network;
pub(crate) mod metadata;

// Debug
pub(crate) mod debug_certificates;

// PerEpoch
pub mod epochs {
    pub(crate) mod certificates;
    pub mod end_checkpoint;
    pub(crate) mod metadata;
    pub(crate) mod proofs;
    pub(crate) mod start_checkpoint;
    mod transaction_hash_per_certificate_index;
}
