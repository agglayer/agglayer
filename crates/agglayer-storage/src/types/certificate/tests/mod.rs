use agglayer_types::{
    aggchain_proof::{Proof, SP1StarkWithContext},
    U256,
};
use pessimistic_proof_test_suite::sample_data;
use sp1_sdk::Prover;

use super::*;
use crate::columns::Codec;

#[rstest::rstest]
#[case(0.into(), [0, 0, 0, 0])]
#[case(100.into(), [0, 0, 0, 100])]
#[case(258.into(), [0, 0, 1, 2])]
#[case(0x12345678.into(), [0x12, 0x34, 0x56, 0x78])]
fn network_id_encoding(#[case] network_id: NetworkId, #[case] expected: [u8; 4]) {
    let encoded = default_bincode_options().serialize(&network_id).unwrap();
    assert_eq!(encoded, expected);
    assert_eq!(network_id.to_u32().to_be_bytes(), expected);
}

fn load_sample_certificate_bytes(filename: &str) -> Vec<u8> {
    let path = std::path::Path::new(env!("CARGO_MANIFEST_DIR"))
        .join("src/types/certificate/tests")
        .join(filename);
    hex::decode(std::fs::read(path).unwrap().trim_ascii()).unwrap()
}

impl From<NetworkId> for NetworkIdV0 {
    fn from(value: NetworkId) -> Self {
        let [b3, b2, b1, b0] = value.to_u32().to_be_bytes();
        assert_eq!(b3, 0);
        NetworkIdV0([b2, b1, b0])
    }
}

impl CertificateV0 {
    fn test0() -> Self {
        Self {
            version: VersionTag,
            network_id: NetworkId::new(55).into(),
            height: 987,
            prev_local_exit_root: Digest([0x01; 32]),
            new_local_exit_root: Digest([0x67; 32]),
            bridge_exits: Vec::new(),
            imported_bridge_exits: Vec::new(),
            signature: Signature::new(
                U256::from_be_bytes([0x78; 32]),
                U256::from_be_bytes([0x9a; 32]),
                false,
            ),
            metadata: Digest([0xa5; 32]),
        }
    }

    fn with_network_id(mut self, network_id: NetworkId) -> Self {
        self.network_id = network_id.into();
        self
    }
}

impl CertificateV1<'static> {
    fn test0() -> Self {
        Self {
            version: VersionTag,
            network_id: NetworkId::new(57),
            height: 987,
            prev_local_exit_root: Digest([0x02; 32]),
            new_local_exit_root: Digest([0x65; 32]),
            bridge_exits: Vec::new().into(),
            imported_bridge_exits: Vec::new().into(),
            aggchain_data: AggchainDataV1::ECDSA {
                signature: Signature::new(
                    U256::from_be_bytes([0x7a; 32]),
                    U256::from_be_bytes([0x9b; 32]),
                    false,
                ),
            },
            metadata: Digest([0xa9; 32]),
            custom_chain_data: Cow::Owned(vec![]),
            l1_info_root: None,
        }
    }

    fn test1() -> Self {
        let (proof, vkey) = {
            let client = sp1_sdk::ProverClient::builder().mock().build();
            let (proving_key, verif_key) = client.setup(pessimistic_proof::ELF);
            let dummy_proof = sp1_sdk::SP1ProofWithPublicValues::create_mock_proof(
                &proving_key,
                sp1_sdk::SP1PublicValues::new(),
                sp1_sdk::SP1ProofMode::Compressed,
                sp1_sdk::SP1_CIRCUIT_VERSION,
            );
            let proof = dummy_proof.proof.try_as_compressed().unwrap();
            (proof, verif_key)
        };

        let proof = Proof::SP1Stark(SP1StarkWithContext {
            proof,
            vkey,
            version: String::from("1.2.3"),
        });

        Self {
            version: VersionTag,
            network_id: NetworkId::new(59),
            height: 987.try_into().unwrap(),
            prev_local_exit_root: Digest([0x03; 32]),
            new_local_exit_root: Digest([0x61; 32]),
            bridge_exits: Vec::new().into(),
            imported_bridge_exits: Vec::new().into(),
            aggchain_data: AggchainDataV1::Generic {
                proof: Cow::Owned(proof),
                aggchain_params: Digest([0x58; 32]),
            },
            metadata: Digest([0xb9; 32]),
            custom_chain_data: Cow::Owned(vec![]),
            l1_info_root: None,
        }
    }
}

impl CertificateV1<'_> {
    fn into_owned(self) -> CertificateV1<'static> {
        let Self {
            version,
            network_id,
            height,
            prev_local_exit_root,
            new_local_exit_root,
            bridge_exits,
            imported_bridge_exits,
            aggchain_data,
            metadata,
            custom_chain_data,
            l1_info_root,
        } = self;

        CertificateV1 {
            version,
            network_id,
            height,
            prev_local_exit_root,
            new_local_exit_root,
            bridge_exits: bridge_exits.into_owned().into(),
            imported_bridge_exits: imported_bridge_exits.into_owned().into(),
            aggchain_data: match aggchain_data {
                AggchainDataV1::ECDSA { signature } => AggchainDataV1::ECDSA { signature },
                AggchainDataV1::Generic {
                    proof,
                    aggchain_params,
                } => AggchainDataV1::Generic {
                    proof: Cow::Owned(proof.into_owned()),
                    aggchain_params,
                },
            },
            metadata,
            custom_chain_data: Cow::Owned(custom_chain_data.into_owned()),
            l1_info_root,
        }
    }
}

#[rstest::rstest]
#[case(CertificateV0::test0(), &[0x00, 0x00, 0x00, 55])]
#[case(CertificateV0::test0().with_network_id(0x123456.into()), &[0x00, 0x12, 0x34, 0x56])]
#[case(CertificateV1::test0(), &[0x01, 0x00, 0x00, 0x00, 57])]
fn encoding_starts_with(#[case] cert: impl Serialize, #[case] start: &[u8]) {
    let bytes = default_bincode_options().serialize(&cert).unwrap();
    assert!(bytes.starts_with(start));
}

#[rstest::rstest]
#[case(CertificateV0::test0())]
#[case(CertificateV1::test0())]
#[case(CertificateV1::test1())]
#[case(CertificateV1::from(&Certificate::new_for_test(74.into(), 998)).into_owned())]
fn encoding_roundtrip_consistent_with_into(#[case] orig: impl Into<Certificate> + Serialize) {
    let bytes = default_bincode_options().serialize(&orig).unwrap();
    let decoded = Certificate::decode(&bytes).unwrap();
    let converted: Certificate = orig.into();

    // This should really compare the certificates directly but that requires adding
    // whole bunch of `Eq` impl to many types.
    assert_eq!(format!("{converted:?}"), format!("{decoded:?}"));
}

#[rstest::rstest]
#[case("n15-cert_h0")]
#[case("n15-cert_h1")]
#[case("n15-cert_h2")]
#[case("n15-cert_h3")]
fn cert_in_v0_format_decodes(#[case] cert_name: &str) {
    let from_json = sample_data::load_certificate(&format!("{cert_name}.json"));

    let bytes = load_sample_certificate_bytes(&format!("encoded_v0-{cert_name}.hex"));
    let from_bytes = Certificate::decode(&bytes).expect("v0 certificate to decode successfully");

    // Again comparing debug output due to lack of `Eq`.
    assert_eq!(format!("{from_bytes:?}"), format!("{from_json:?}"));
}

#[test]
fn bad_format() {
    const NEXT_VERSION: u8 = 2;

    assert!(matches!(
        Certificate::decode(&[]).unwrap_err(),
        CodecError::CertificateEmpty
    ));

    for v in 0..NEXT_VERSION {
        assert!(matches!(
            Certificate::decode(&[v]).unwrap_err(),
            CodecError::Serialization(_)
        ));
    }

    for v in NEXT_VERSION..=u8::MAX {
        match Certificate::decode(&[v]).unwrap_err() {
            CodecError::BadCertificateVersion { version } => assert_eq!(version, v),
            err => panic!("Unexpected error: {err:?}"),
        }
    }
}
