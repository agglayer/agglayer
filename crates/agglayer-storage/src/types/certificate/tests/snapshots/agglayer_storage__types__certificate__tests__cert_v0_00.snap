---
source: crates/agglayer-storage/src/types/certificate/tests/mod.rs
expression: bytes
snapshot_kind: text
---
0x0000003700000000000003db01010101010101010101010101010101010101010101010101010101010101016767676767676767676767676767676767676767676767676767676767676767000000000000000000000000000000000000000000000020787878787878787878787878787878787878787878787878787878787878787800000000000000209a9a9a9a9a9a9a9a9a9a9a9a9a9a9a9a9a9a9a9a9a9a9a9a9a9a9a9a9a9a9a9a00a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5
